### ISSUE TITLE
Transfer learning on multistep models preserving decoder weights

### ISSUE THREAD

**Issue by MicheleCattaneo** [ecmwf/anemoi-core#783]
## Main Idea
With the introduction of multistep models (https://github.com/ecmwf/anemoi-core/pull/636), it is possible to configure the model to forecast more than a single step ahead.
Currently, when doing transfer learning, any trainable tensor that does not match the name and shape of the source checkpoint is ignored when the loading the pre-trained weights [(See here)](https://github.com/ecmwf/anemoi-core/blob/a161fe20dfae159b59bbe4d507b55abb3653d592/training/src/anemoi/training/utils/checkpoint.py#L82).

However, in a transfer learning setting between two multistep models that forecast a different number of steps ahead there can be an overlap for certain lead times.
Because all timesteps are jointly forecasted from the same latent vector using a fully connected layer, parts of trained weights in this layer can be preserved when loading the model, without forgetting what was previously learned for the common lead times.

Original idea suggested by: @dietervdb-meteo 
## Example

A first model is trained to forecast $[+3h, +6h]$. The picture below represents the last linear layer in the decoder, that maps a latent vector $h$ into the output vector $x$ of shape $2\cdot n_{vars}$ (all the variables for two time steps).

<img width="420" alt="Image" src="https://github.com/user-attachments/assets/df267a09-d33a-436c-8112-ea3bfaba3a81" />

A new model that needs to forecast $[+1h, +2h,+3h,+4h,+5h,+6h]$ could re-utilize the weights associated with $[+3h, +6h]$ from the previous model as shown below and output a vector $x$ of shape $6\cdot n_{vars}$

<img width="800" alt="Image" src="https://github.com/user-attachments/assets/a2c85fff-c2ee-4f31-b915-909af66d28d8" />

This model would start its training procedure having retained all the skill of the first model for leads times $[+3h, +6h]$.

## Proposal

I propose to extend the transfer learning logic in `anemoi-training` to recognize common lead times when doing transfer learning between two models and preserve the associated weights when loading the new model.

Weights associated with lead times that are not in common between the two models could:
- Be randomly initialized 
- Be the same as the closest common lead time (e.g: in the case above $W_4$ would be set to be $W_3$, and the model would initially forecast the same value for $+4h$ as for $+3h$.

## Impact

- It could be possible to transfer old single step ahead checkpoints into multistep models
- It could facilitate the training of multistep models following a curriculum, where models are first trained to forecast a smaller number of lead times, increased over time.

---
I am willing to contribute and I am about to open a draft PR with some initial implementation.

**Comment by ssmmnn11** [comment]
Is it just about the final linear layer? then what could also be done is to have a backward mapper that supports multiple final decoder layers, during pre-training you would have a [module list of linear layers](https://docs.pytorch.org/docs/stable/generated/torch.nn.ModuleList.html) or [module dict](https://docs.pytorch.org/docs/stable/generated/torch.nn.ModuleDict.html) instead of a single final layer, then for pre-training you would only have one or two, and then you extend this list for fine-tuning on more steps. Finally you get the output via torch.stack(...)

**Comment by ssmmnn11** [comment]
Good point, we would also need a loop for the input steps and an associated ModuleDict. I would not worry about efficiency, these layers contribution to the total model runtime is tiny - and compared to communication etc. would be negligible. Alternatively you could use the same layer(s) to embed and extract the steps etc.

**Comment by MicheleCattaneo** [comment]
I have started a small implementation in a branch (no PR for now, I am still not sure if it's best to wait for the #636 to be merged) where I extract the submatrices related to the various lead times and transfer them to the new layer. This is still all on standalone layers, not yet in the whole model logic. See:
- [implementation](https://github.com/ecmwf/anemoi-core/blob/6dd28b4e6576e5aa1ceccaa078d50ab633af2ade/training/src/anemoi/training/utils/checkpoint.py#L78)
- [usage (tests)](https://github.com/ecmwf/anemoi-core/blob/6dd28b4e6576e5aa1ceccaa078d50ab633af2ade/training/tests/unit/train/test_transfer_learning.py#L31)

Would this be what you meant as "embed and extract the steps"?

**Comment by ssmmnn11** [comment]
I was thinking, instead of doing something like this:

einops.rearrange(x, "batch time ensemble grid vars -> (batch ensemble grid) (time vars)"),

keep the time dimension of the input (and output steps), and then doing a loop over the timesteps, for example:

out = []
for i in range(steps):
    out.append(lin_layers[i](x[:, i, ...]))
out = torch.stack(...)

or you add instead etc. 

then this can go through the the encoder ... decoder and you can compute the loss for each step (you might need different normalisation / de-normalisation for e.g. tendency prediction and you might want to weight different steps in a different way).

You end up with a model where you can just add steps (or variables for that matter).

Or we could change the encoder that it receives a list of tensors, and it just loops over it and embeds it like above. You could just concatenate a time step index to each tensor.
